crates/orchestrator/src/queue/job_queue.rs

```diff
@@ -52,60 +83,134 @@ where
+                            .map_err(|e| ConsumptionError::Other(OtherError::from(e)))?,
+                        Err(e) => {
+                            log::error!("Failed to handle job with id {:?}. Error: {:?}", job_message.id, e);
+
+                            // if the queue as a retry logic at the source, it will be attempted
+                            // after the nack
+                            match d.nack().await {
+                                Ok(_) => Err(ConsumptionError::FailedToHandleJob {
+                                    job_id: job_message.id,
+                                    error_msg: "Job handling failed, message nack-ed".to_string(),
+                                })?,
+                                Err(delivery_nack_error) => Err(ConsumptionError::FailedToHandleJob {
+                                    job_id: job_message.id,
+                                    error_msg: delivery_nack_error.0.to_string(),
+                                })?,
+                            }
+                        }
+                    };
+                }
+                None => return Ok(()),
+            }
         }
-    };
-    let job_message: Option<JobQueueMessage> = delivery
-        .payload_serde_json()
-        .wrap_err("Payload Serde Error")
-        .map_err(|e| ConsumptionError::Other(OtherError::from(e)))?;
-
-    match job_message {
-        Some(job_message) => {
-            log::info!("Handling job with id {:?} for queue {:?}", job_message.id, queue);
-            match handler(job_message.id).await {
-                Ok(_) => delivery
-                    .ack()
-                    .await
-                    .map_err(|(e, _)| e)
-                    .wrap_err("Queue Error")
-                    .map_err(|e| ConsumptionError::Other(OtherError::from(e)))?,
-                Err(e) => {
-                    log::error!("Failed to handle job with id {:?}. Error: {:?}", job_message.id, e);
-
-                    // if the queue as a retry logic at the source, it will be attempted
-                    // after the nack
-                    match delivery.nack().await {
-                        Ok(_) => Err(ConsumptionError::FailedToHandleJob {
-                            job_id: job_message.id,
-                            error_msg: "Job handling failed, message nack-ed".to_string(),
-                        })?,
-                        Err(delivery_nack_error) => Err(ConsumptionError::FailedToHandleJob {
-                            job_id: job_message.id,
-                            error_msg: delivery_nack_error.0.to_string(),
-                        })?,
-                    }
+        DeliveryReturnType::NoMessage => return Ok(()),
+    }
+
+    Ok(())
+}
+
+/// Function to consume the message from the worker trigger queues and spawn the worker
+/// for respective message received.
+pub async fn consume_worker_trigger_messages_from_queue<F, Fut>(
+    queue: String,
+    handler: F,
```

Comment:
A design question I have: why is the handler async if you're going to `await` it anyway? It seems like you're not taking advantage of the asynchronicity here by running multiple workers in parallel. Is it something you'll support later? And/or do you need the handler to be async because it calls async functions itself?

Reply:
Here technically all the queues are consuming the messages in parellel only or I didn't understand your question. 

Here I am adding the consumers : 

```rust
macro_rules! spawn_consumer {
    ($queue_type :expr, $handler : expr, $consume_function: expr) => {
        tokio::spawn(async move {
            loop {
                match $consume_function($queue_type, $handler).await {
                    Ok(_) => {}
                    Err(e) => log::error!("Failed to consume from queue {:?}. Error: {:?}", $queue_type, e),
                }
                sleep(Duration::from_secs(1)).await;
            }
        });
    };
}

pub async fn init_consumers() -> Result<(), JobError> {
    spawn_consumer!(JOB_PROCESSING_QUEUE.to_string(), process_job, consume_job_from_queue);
    spawn_consumer!(JOB_VERIFICATION_QUEUE.to_string(), verify_job, consume_job_from_queue);
    spawn_consumer!(JOB_HANDLE_FAILURE_QUEUE.to_string(), handle_job_failure, consume_job_from_queue);
    spawn_consumer!(WORKER_TRIGGER_QUEUE.to_string(), spawn_worker, consume_worker_trigger_messages_from_queue);
    Ok(())
}
```

Can you explain more on this ? that would be helpful. 
Thanks

Reply:
Yes, I saw that you spawned multiple consumers, one per queue. My remark is that each queue consumer is sequential, i.e. all messages coming from one queue are treated one after the other here. There's no inherent problem with that, I was just wondering if this was the design you were going for or if you wanted to treat messages from each queue in parallel in the future as well.

Reply:
Good point I will create an issue for this and we will process all the queue messages in parallel